src/Featurizers/LagLeadOperatorFeaturizer.h

```diff
@@ -0,0 +1,234 @@
+    // |
+    // ----------------------------------------------------------------------
+    static_assert(Components::Details::IsTuple<InputTupleT>::value, "'InputTupleT' must be a tuple");
+
+    using BaseType =
+        Transformer<
+            InputTupleT,
+            Microsoft::Featurizer::ColMajMatrix<std::double_t>
+        >;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    LagLeadOperatorTransformer(std::uint32_t horizon, std::vector<std::int64_t> offsets);
+    LagLeadOperatorTransformer(Archive &ar);
+
+    ~LagLeadOperatorTransformer(void) override = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(LagLeadOperatorTransformer);
+
+    bool operator==(LagLeadOperatorTransformer const &other) const;
+    bool operator!=(LagLeadOperatorTransformer const &other) const;
+
+    void save(Archive &ar) const override;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Types
+    // |
+    // ----------------------------------------------------------------------
+    using GrainType  = std::vector<std::string>;
+    // TODO: change target type to general types
+    using TargetType = std::double_t;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    Components::CircularBuffer<TargetType>  _buffer;
+    std::int64_t const                      _max_future_offset;
+    std::int64_t const                      _max_prior_offset;
+    std::uint32_t const                     _horizon;
+    std::vector<std::int64_t> const         _offsets;
+    std::int64_t                            _index;
+    std::uint32_t                           _numPending;
+    
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    // MSVC runs into problems when the declaration and definition are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        if (_max_future_offset != 0 && _index <= _max_future_offset + _horizon - 1) {
+            ++_index;
+            ++_numPending;
+            _buffer.push(std::get<1>(input));
+            return;
+        }
+        Microsoft::Featurizer::ColMajMatrix<std::double_t> ret(_offsets.size(), _horizon);
+        for (std::uint32_t col = 0; col < _horizon; ++col) {
+            for (size_t row = 0; row < _offsets.size(); ++row) {
+                if ((_index - _numPending - (_horizon - 1 - col) + _offsets[row] < 0)) {
+                    ret(static_cast<long long>(row), col) = std::nan("");
```

Comment:
>"" [](start = 69, length = 2)

What is the meaning of this, passing an empty const char* pointer to a function that expects integral types?

Reply:
It's creating a quiet NaN and the method requires a const char * parameter which I think I can leave as empty in this case.